# **DATA EXPLORATION AND UNDERSTANDING**

## 1. Business Understanding

`INPUT VARIABLES`: numerical and categorical <br>
`OUPUT VARIABLE`: numeric (continuous variable) <br>

__GOAL__: Predict the price of cars

`carID` : An attribute that contains an identifier for each car <br>
`Brand` : The car's main brand number <br>
`Model` : The car model <br>
`year`: The year of Registration of the Car <br>
`mileage` : The total reported distance travelled by the car (in miles) <br>
`tax` : The amount of road tax that, in 2020, was applicable to the car in question <br>
`fuelType`: Type of Fuel used by the car <br>
`mpg`: Average Miles per Gallon <br>
`engineSize`: Size of Engine in liters (Cubic Decimeters) <br>
`paintQuality%`: The mechanic's assessment of the cars' overall paint quality and hull integrity (filled by the mechanic during evaluation) <br>
`previousOwners`: Number of previous registered owners of the vehicle <br>
`hasDamage`: Boolean marker filled by the seller at the time of registration stating whether the car is damaged or not <br>
`price`: The car's price when purchased by Cars 4 You <br>

## 2. Imports and Read Data

In [2751]:
# importing the libraries
import pandas as pd
import numpy as np
from datetime import timedelta
from matplotlib import pyplot as plt
import math
import seaborn as sns
import re
import json
from collections import Counter

In [2752]:
# setting the options
#pd.set_option('display.max_columns', None)
#pd.set_option('display.max_colwidth', None)
#pd.set_option('display.max_rows', None)
#np.set_printoptions(threshold=np.inf)
#plt.style.use('seaborn-v0_8-dark')

In [2753]:
# importing the data
# df_training = pd.read_csv('C:/Users/calco/OneDrive/Ambiente de Trabalho/nova ims/1 Semestre/ML/project_data/train.csv', sep=',')
# df_test = pd.read_csv('C:/Users/calco/OneDrive/Ambiente de Trabalho/nova ims/1 Semestre/ML/project_data/test.csv', sep=',')

df_training = pd.read_csv('project_data/train.csv', sep=',')
df_test = pd.read_csv('project_data/test.csv', sep=',')


## 3. Macro-Inspection

In [2754]:
#Check the shape of the dataset
df_training.shape

(75973, 14)

In [2755]:
# check the first rows of the dataset
df_training.head()

,carID,Brand,model,year,price,transmission,mileage,fuelType,tax,mpg,engineSize,paintQuality%,previousOwners,hasDamage
0,69512,VW,Golf,2016.0,22290,Semi-Auto,28421.0,Petrol,NaN,11.417268,2.0,63.0,4.000000,0.0
1,53000,Toyota,Yaris,2019.0,13790,Manual,4589.0,Petrol,145.0,47.900000,1.5,50.0,1.000000,0.0
2,6366,Audi,Q2,2019.0,24990,Semi-Auto,3624.0,Petrol,145.0,40.900000,1.5,56.0,4.000000,0.0
3,29021,Ford,FIESTA,2018.0,12500,anual,9102.0,Petrol,145.0,65.700000,1.0,50.0,-2.340306,0.0
4,10062,BMW,2 Series,2019.0,22995,Manual,1000.0,Petrol,145.0,42.800000,1.5,97.0,3.000000,0.0


In [2756]:
# check if there are any aggregation rows at the end of the dataset
df_training.tail()

,carID,Brand,model,year,price,transmission,mileage,fuelType,tax,mpg,engineSize,paintQuality%,previousOwners,hasDamage
75968,37194,Mercedes,C Class,2015.0,13498,Manual,14480.0,etrol,125.0,53.3,2.0,78.0,0.0,0.0
75969,6265,Audi,Q3,2013.0,12495,Semi-Auto,52134.0,Diesel,200.0,47.9,2.0,38.0,2.0,0.0
75970,54886,Toyota,Aygo,2017.0,8399,Automatic,11304.0,Petrol,145.0,67.0,1.0,57.0,3.0,0.0
75971,860,Audi,Q3,2015.0,12990,Manual,69072.0,iesel,125.0,60.1,2.0,74.0,2.0,0.0
75972,15795,Ford,Fiesta,2018.0,10495,Manual,16709.0,Petro,145.0,64.2,1.1,38.0,1.0,0.0


In [2757]:
# check the datatypes and null counts
df_training.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 75973 entries, 0 to 75972
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   carID           75973 non-null  int64  
 1   Brand           74452 non-null  object 
 2   model           74456 non-null  object 
 3   year            74482 non-null  float64
 4   price           75973 non-null  int64  
 5   transmission    74451 non-null  object 
 6   mileage         74510 non-null  float64
 7   fuelType        74462 non-null  object 
 8   tax             68069 non-null  float64
 9   mpg             68047 non-null  float64
 10  engineSize      74457 non-null  float64
 11  paintQuality%   74449 non-null  float64
 12  previousOwners  74423 non-null  float64
 13  hasDamage       74425 non-null  float64
dtypes: float64(8), int64(2), object(4)
memory usage: 8.1+ MB


Only **carID** and **price** don't have nan values

In [2758]:
#Check the data types of the features
df_training.dtypes

carID               int64
Brand              object
model              object
year              float64
price               int64
transmission       object
mileage           float64
fuelType           object
tax               float64
mpg               float64
engineSize        float64
paintQuality%     float64
previousOwners    float64
hasDamage         float64
dtype: object

**Data type analysis:**

Features that should be booleans:
- **'hasDamage'**



In [2759]:
df_training.describe(include='all').T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
carID,75973.0,NaN,NaN,NaN,37986.0,21931.660338,0.0,18993.0,37986.0,56979.0,75972.0
Brand,74452,72,Ford,14808,NaN,NaN,NaN,NaN,NaN,NaN,NaN
model,74456,735,Focus,6353,NaN,NaN,NaN,NaN,NaN,NaN,NaN
year,74482.0,NaN,NaN,NaN,2017.096611,2.208704,1970.0,2016.0,2017.0,2019.0,2024.121759
price,75973.0,NaN,NaN,NaN,16881.889553,9736.926322,450.0,10200.0,14699.0,20950.0,159999.0
transmission,74451,40,Manual,38050,NaN,NaN,NaN,NaN,NaN,NaN,NaN
mileage,74510.0,NaN,NaN,NaN,23004.184088,22129.788366,-58540.574478,7423.25,17300.0,32427.5,323000.0
fuelType,74462,34,Petrol,37995,NaN,NaN,NaN,NaN,NaN,NaN,NaN
tax,68069.0,NaN,NaN,NaN,120.329078,65.521176,-91.12163,125.0,145.0,145.0,580.0
mpg,68047.0,NaN,NaN,NaN,55.152666,16.497837,-43.421768,46.3,54.3,62.8,470.8


## 4. Identifier Variable

This feature is the unique identifier of each claim - we will analyse this column in more depth in an attempt to assign it as the index of our dataframe.

In [2760]:
df_training[df_training['carID'].duplicated(keep=False)]

,carID,Brand,model,year,price,transmission,mileage,fuelType,tax,mpg,engineSize,paintQuality%,previousOwners,hasDamage


Upon review, there are no duplicated 'carID' values in the dataset. </br>
As we saw previously, this column has no null values, so we can set it as the dataframe index.

In [2761]:
df_training = df_training.set_index('carID').rename_axis(None)
df_training

,Brand,model,year,price,transmission,mileage,fuelType,tax,mpg,engineSize,paintQuality%,previousOwners,hasDamage
69512,VW,Golf,2016.0,22290,Semi-Auto,28421.0,Petrol,NaN,11.417268,2.0,63.0,4.000000,0.0
53000,Toyota,Yaris,2019.0,13790,Manual,4589.0,Petrol,145.0,47.900000,1.5,50.0,1.000000,0.0
6366,Audi,Q2,2019.0,24990,Semi-Auto,3624.0,Petrol,145.0,40.900000,1.5,56.0,4.000000,0.0
29021,Ford,FIESTA,2018.0,12500,anual,9102.0,Petrol,145.0,65.700000,1.0,50.0,-2.340306,0.0
10062,BMW,2 Series,2019.0,22995,Manual,1000.0,Petrol,145.0,42.800000,1.5,97.0,3.000000,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
37194,Mercedes,C Class,2015.0,13498,Manual,14480.0,etrol,125.0,53.300000,2.0,78.0,0.000000,0.0
6265,Audi,Q3,2013.0,12495,Semi-Auto,52134.0,Diesel,200.0,47.900000,2.0,38.0,2.000000,0.0
54886,Toyota,Aygo,2017.0,8399,Automatic,11304.0,Petrol,145.0,67.000000,1.0,57.0,3.000000,0.0
860,Audi,Q3,2015.0,12990,Manual,69072.0,iesel,125.0,60.100000,2.0,74.0,2.000000,0.0


In [2762]:
df_test = df_test.set_index('carID').rename_axis(None)

## 5. Missing Values

Has we saw previously only **carId** and the **price** don't have nana values

In [2763]:
# calculate the number of NaNs for each column
nan_counts = df_training.isna().sum()

# get the total number of rows
total_rows = df_training.shape[0]

# calculate the percentage of NaN values for each column
percentage_nans = round((nan_counts / total_rows) * 100,2)

# format the percentages with '%' sign
#percentage_nans = percentage_nans.apply(lambda x: f"{x:.2f}%")

# combine all the information into a DataFrame
nan_summary = pd.DataFrame({
    'NaN Count': nan_counts
    , 'Total Values': [total_rows] * len(nan_counts)
    , 'Percentage NaN(%)': percentage_nans
})

display(nan_summary)

,NaN Count,Total Values,Percentage NaN(%)
Brand,1521,75973,2.00
model,1517,75973,2.00
year,1491,75973,1.96
price,0,75973,0.00
transmission,1522,75973,2.00
mileage,1463,75973,1.93
fuelType,1511,75973,1.99
tax,7904,75973,10.40
mpg,7926,75973,10.43
engineSize,1516,75973,2.00


By analyzing the output above, we can conclude:
- **'tax'** and **'mpg'**: approximately 10% of the data is missing for these features. 
- All other features have a minimal proportion of missing values, around 2%, which is not significant.

In [2764]:
#check if there is a row with all values are NaN
df_training[df_training.isna().all(axis=1)]

,Brand,model,year,price,transmission,mileage,fuelType,tax,mpg,engineSize,paintQuality%,previousOwners,hasDamage


In [2765]:
nan_ratio = df_training.isna().mean(axis=1).sort_values()
nan_ratio

54886    0.000000
6265     0.000000
37194    0.000000
60263    0.000000
16023    0.000000
           ...   
27486    0.307692
20150    0.307692
44389    0.384615
27265    0.384615
29984    0.384615
Length: 75973, dtype: float64

## 6. Unique Values

In [2766]:
df_training.nunique() #to see if there is more than 1 value in each feature

Brand                72
model               735
year                 44
price             11048
transmission         40
mileage           34808
fuelType             34
tax                  65
mpg                 217
engineSize           58
paintQuality%        88
previousOwners       23
hasDamage             1
dtype: int64

The feature **hasDamage** only has one type of value, meaning is a irrelavant feature

In [2767]:
# df_training.drop(columns='hasDamage', inplace = True)
# df_training

In [2768]:
#df_test.drop(columns='hasDamage', inplace=True)

## 7.Data Types
We will check each column for mixed data types and convert features to the appropriate types.

In [2769]:
# Transformar nomes das colunas
df_training.columns = [col.capitalize() for col in df_training.columns]
df_test.columns = [col.capitalize() for col in df_test.columns]
print(df_training.columns)

Index(['Brand', 'Model', 'Year', 'Price', 'Transmission', 'Mileage',
       'Fueltype', 'Tax', 'Mpg', 'Enginesize', 'Paintquality%',
       'Previousowners', 'Hasdamage'],
      dtype='object')


In [2770]:
df_training.rename(columns={
    'Fueltype': 'FuelType',
    'Enginesize': 'EngineSize',
    'Paintquality%': 'PaintQuality%',
    'Previousowners': 'PreviousOwners',
    'Mpg': 'MPG'
}, inplace=True)

df_test.rename(columns={
    'Fueltype': 'FuelType',
    'Enginesize': 'EngineSize',
    'Paintquality%': 'PaintQuality%',
    'Previousowners': 'PreviousOwners',
    'Mpg': 'MPG'
}, inplace=True)


In [2771]:
df_training.dtypes

Brand              object
Model              object
Year              float64
Price               int64
Transmission       object
Mileage           float64
FuelType           object
Tax               float64
MPG               float64
EngineSize        float64
PaintQuality%     float64
PreviousOwners    float64
Hasdamage         float64
dtype: object

In [2772]:
#check if there is any feature who has more than 1 dtype
def check_mixed_types(column):
    # if there are more than one unique data types (exclude NaN values, as these are considered floats)
    return len(set(column[pd.notna(column)].apply(type))) > 1

In [2773]:
mixed_type_columns = [col for col in df_training.columns if check_mixed_types(df_training[col])]
mixed_type_columns_test = [col for col in df_test.columns if check_mixed_types(df_test[col])]

print(mixed_type_columns)
print(mixed_type_columns_test)

[]
[]


In [2774]:
df_test.columns

Index(['Brand', 'Model', 'Year', 'Transmission', 'Mileage', 'FuelType', 'Tax',
       'MPG', 'EngineSize', 'PaintQuality%', 'PreviousOwners', 'Hasdamage'],
      dtype='object')

## 8. Categorical Features

In [2775]:
# df_copy=df.copy()
# df_test_copy = df_test.copy()

In [2776]:
numerical_columns = df_training.select_dtypes(include=['number']).columns
categorical_columns = df_test.select_dtypes(include=['object']).columns
categorical_columns

Index(['Brand', 'Model', 'Transmission', 'FuelType'], dtype='object')

In [2777]:
for col in categorical_columns:
    print(f"\n--- {col} ---")
    print(df_training[col].unique())


--- Brand ---
['VW' 'Toyota' 'Audi' 'Ford' 'BMW' 'Skoda' 'Opel' 'Mercedes' 'FOR'
 'mercedes' 'Hyundai' 'w' 'ord' 'MW' 'bmw' nan 'yundai' 'BM' 'Toyot' 'udi'
 'Ope' 'AUDI' 'V' 'opel' 'pel' 'For' 'pe' 'Mercede' 'audi' 'MERCEDES'
 'OPEL' 'koda' 'FORD' 'Hyunda' 'W' 'Aud' 'vw' 'hyundai' 'skoda' 'ford'
 'TOYOTA' 'ercedes' 'oyota' 'toyota' 'SKODA' 'Skod' 'HYUNDAI' 'kod' 'v'
 'for' 'SKOD' 'aud' 'KODA' 'PEL' 'yunda' 'or' 'UDI' 'OYOTA' 'HYUNDA' 'mw'
 'OPE' 'mercede' 'ERCEDES' 'ercede' 'TOYOT' 'MERCEDE' 'ORD' 'ud' 'ope'
 'AUD' 'hyunda' 'skod' 'toyot']

--- Model ---
[' Golf' ' Yaris' ' Q2' ' FIESTA' ' 2 Series' '3 Series' ' A3' ' Octavia'
 ' Passat' ' Focus' ' Insignia' ' A Clas' ' Q3' ' Fabia' ' A Class' ' Ka+'
 ' 3 Series' ' GLC Class' ' I30' ' C Class' ' Polo' ' E Class' ' C Clas'
 ' Q5' ' Up' ' Fiesta' ' C-HR' ' Mokka X' ' Corsa' ' Astra' ' TT'
 ' 5 Series' ' Aygo' ' 4 Series' ' SLK' ' Viva' ' T-Roc' 'Focus'
 ' EcoSport' ' Tucson' ' EcoSpor' nan ' X-CLASS' ' CL Class' ' IX20'
 ' i20' ' Rapid'

In **brands**,**model**,, **transmission**, **fuel type**  we can see that are weird values and also names that are in capitall letter and low letter

In [2778]:
# lets first change every value to capital letter
df_training['Brand']=df_training['Brand'].str.upper()

#now lets change the assumning filling error, to the full name of the brand
brand_lists = {
    'MERCEDES': ['MERCEDE', 'ERCEDES', 'ERCEDE'],
    'VW': ['V', 'MW', 'W'],
    'AUDI': ['UDI', 'AUD', 'UD'],
    'FORD': ['FOR', 'ORD', 'OR'],
    'BMW': ['MW', 'BM'],
    'HYUNDAI': ['YUNDAI', 'YUNDA', 'HYUNDA'],
    'SKODA': ['KODA', 'SKOD', 'KOD'],
    'TOYOTA': ['TOYOT','OYOTA',],
    'OPEL': ['OPE','PEL','PE']
    
}

# Loop through the dictionary and replace
for standard, variations in brand_lists.items():
    df_training['Brand'] = df_training['Brand'].replace(variations, standard)
    df_test['Brand'] = df_test['Brand'].replace(variations, standard)
    
df_training['Brand'].unique()

array(['VW', 'TOYOTA', 'AUDI', 'FORD', 'BMW', 'SKODA', 'OPEL', 'MERCEDES',
       'HYUNDAI', nan], dtype=object)

In [2779]:
# lets change every value to capital letter
df_training['Model']=df_training['Model'].str.upper()
#there are also felling errors like '2 SERIES' and ' 2 Series'-> extra space problem
df_training['Model']=df_training['Model'].str.strip()

corrections = {
    'ARTEO':'ARTEON',#N
    'YET':'YETI',#I
    'AMARO':'AMAROK',#K
    'RAPI':'RAPID',#D
    'GOL':'GOLF',#F
    'T-RO':'T-ROC',#C
    'GALAX':'GALAXY',#Y
    'SUPER':'SUPERB',#B
    'TOUARE':'TOUAREG',#G
    'U':'UP',#P
    'I80':'I800',
    'T':'TT',
    'VIV':'VIVARO',
    'ADA':'ADAM'

}

df_training['Model'] = df_training['Model'].replace(corrections)
  
hyun_brand=df_training['Brand']=='HYUNDAI'
i_start = df_training['Model'].str.startswith('I', na=False)
len_2 = df_training['Model'].str.len() == 2
filter= hyun_brand & i_start & len_2

def clean_model(val):
    if not isinstance(val, str):
        return val
    elif  val == 'VERSO-SO' or val=='CALIFORNIAE':
        return val[:-1]
    if val.endswith(('CLAS','SERIE','-CROS','FOCU','YARI','AURI')):
        return val + 'S'
    if val.endswith(('-MA','HILU')):
        return val + 'X'
    if val.endswith(('CRUISE','VELOSTE','-H','TOURE','OUTDOO','ROOMSTE')) :
        return val + 'R'
    if val.endswith(('VERS','SCIROCC','CITIG','MONDE','AYG','POL')) :
        return val + 'O'
    if val.endswith(('EDG','CARAVELL','LIF','BEETL','ALLSPAC','SANTA F')) :
        return val + 'E'
    if val.endswith(('CUSTO',)):
        return val + 'M'
    if val.endswith(('CONNEC','ECOSPOR','PASSA',)):
        return val + 'T'   
    if val.endswith(('FIEST', 'CORS','ASTR','INSIGNI','ZAFIR','OCTAVI','SCAL','KON','MOKK','KUG','FABI','MERIV','COROLL')):
        return val + 'A' 
    if val.endswith(('TIGUA', 'TUCSO','TOURA','SHARA')):
        return val + 'N' 
    if val.endswith(('KARO', 'KAMI','IONI','KODIA')):
        return val + 'Q' 
    if val.endswith(('220', '230', '200')):
        return 'C'+ val 
    return val

df_training['Model'] = df_training['Model'].apply(clean_model)
df_training.loc[filter, 'Model'] = df_training.loc[filter, 'Model'] + '0'

df_training['Model'].unique()

array(['GOLF', 'YARIS', 'Q2', 'FIESTA', '2 SERIES', '3 SERIES', 'A3',
       'OCTAVIA', 'PASSAT', 'FOCUS', 'INSIGNIA', 'A CLASS', 'Q3', 'FABIA',
       'KA+', 'GLC CLASS', 'I30', 'C CLASS', 'POLO', 'E CLASS', 'Q5',
       'UP', 'C-HR', 'MOKKA X', 'CORSA', 'ASTRA', 'TT', '5 SERIES',
       'AYGO', '4 SERIES', 'SLK', 'VIVA', 'T-ROC', 'ECOSPORT', 'TUCSON',
       nan, 'X-CLASS', 'CL CLASS', 'IX20', 'I20', 'RAPID', 'A1', 'AURIS',
       'SHARAN', 'ADAM', 'X3', 'A8', 'GLS CLASS', 'B-MAX', 'A4', 'KONA',
       'I10', 'MOKKA', 'S-MAX', 'X2', 'CROSSLAND X', 'TIGUAN', 'A5',
       'GLE CLASS', 'ZAFIRA', 'IONIQ', 'A6', 'MONDEO', 'YETI OUTDOOR',
       'X1', 'SCALA', 'S CLASS', '1 SERIES', 'KAMIQ', 'KUGA',
       'TOURNEO CONNECT', 'Q7', 'GLA CLASS', 'ARTEON', 'SL CLASS',
       'SANTA FE', 'GRANDLAND X', 'I800', 'RAV4', 'TOURAN', 'CITIGO',
       'ROOMSTER', 'PRIUS', 'COROLLA', 'B CLASS', 'Q', 'KODIAQ',
       'V CLASS', 'CADDY MAXI LIFE', 'SUPERB', 'GETZ', 'COMBO LIFE',
       'BEETLE', 'GALAXY

In [2780]:
#for the test

df_test['Model']=df_test['Model'].str.upper()
#there are also felling errors like '2 SERIES' and ' 2 Series'-> extra space problem
df_test['Model']=df_test['Model'].str.strip()

df_test['Model'] = df_test['Model'].replace(corrections)
  
hyun_brand=df_test['Brand']=='HYUNDAI'
i_start = df_test['Model'].str.startswith('I', na=False)
len_2 = df_test['Model'].str.len() == 2
filter= hyun_brand & i_start & len_2

df_test['Model'] = df_test['Model'].apply(clean_model)
df_test.loc[filter, 'Model'] = df_test.loc[filter, 'Model'] + '0'


In [2781]:
df_training['FuelType']=df_training['FuelType'].str.upper()
df_training['FuelType']=df_training['FuelType'].str.strip()

FuelType_lists = { 
    'PETROL': ['ETROL', 'ETRO','PETRO'],
    'DIESEL': ['DIESE', 'IESE', 'IESEL'],
    'HYBRID': ['YBRID', 'HYBRI','YBRI'],
    'OTHER': ['THER','OTHE']
}

for new, old in FuelType_lists.items():
    df_training['FuelType']=df_training['FuelType'].replace(old, new)
    df_test['FuelType']=df_test['FuelType'].replace(old, new)

df_training['FuelType'].unique()

array(['PETROL', 'DIESEL', 'HYBRID', nan, 'OTHER', 'ELECTRIC'],
      dtype=object)

In [2782]:
#Other doesn't make much sense so I will turn all the Other to unknown values

df_training.loc[df_training['FuelType']=='OTHER','FuelType']='UNKNOWN'

df_training['FuelType'].unique()

array(['PETROL', 'DIESEL', 'HYBRID', nan, 'UNKNOWN', 'ELECTRIC'],
      dtype=object)

In [2783]:
df_training['Transmission']=df_training['Transmission'].str.upper()
df_training['Transmission']=df_training['Transmission'].str.strip()

transmission_list = {
    'SEMI-AUTOMATIC': ['SEMI-AUT','SEMI-AUT','EMI-AUTO','SEMI-AUTO','EMI-AUT'],
    'MANUAL': ['ANUAL','MANUA','ANUA'],
    'AUTOMATIC': ['AUTOMATI', 'UTOMATIC','UTOMATI'],
    'UNKNOWN': ['NKNOWN', 'NKNOW', 'UNKNOW']
}

for word, value in transmission_list.items():
    df_training['Transmission'] = df_training['Transmission'].replace(value, word)
    df_test['Transmission'] = df_test['Transmission'].replace(value, word)

df_training['Transmission'].unique()

array(['SEMI-AUTOMATIC', 'MANUAL', 'AUTOMATIC', nan, 'UNKNOWN', 'OTHER'],
      dtype=object)

In [2784]:
#Same with transmission_list other doesn't make much sense so I will turn all the Other to unknown values

df_training.loc[df_training['Transmission']=='OTHER','Transmission']='UNKNOWN'

df_training['Transmission'].unique()

array(['SEMI-AUTOMATIC', 'MANUAL', 'AUTOMATIC', nan, 'UNKNOWN'],
      dtype=object)

In [2785]:
for col in categorical_columns:
    print(f"\n--- {col} ---")
    print(df_training[col].unique())


--- Brand ---
['VW' 'TOYOTA' 'AUDI' 'FORD' 'BMW' 'SKODA' 'OPEL' 'MERCEDES' 'HYUNDAI' nan]

--- Model ---
['GOLF' 'YARIS' 'Q2' 'FIESTA' '2 SERIES' '3 SERIES' 'A3' 'OCTAVIA'
 'PASSAT' 'FOCUS' 'INSIGNIA' 'A CLASS' 'Q3' 'FABIA' 'KA+' 'GLC CLASS'
 'I30' 'C CLASS' 'POLO' 'E CLASS' 'Q5' 'UP' 'C-HR' 'MOKKA X' 'CORSA'
 'ASTRA' 'TT' '5 SERIES' 'AYGO' '4 SERIES' 'SLK' 'VIVA' 'T-ROC' 'ECOSPORT'
 'TUCSON' nan 'X-CLASS' 'CL CLASS' 'IX20' 'I20' 'RAPID' 'A1' 'AURIS'
 'SHARAN' 'ADAM' 'X3' 'A8' 'GLS CLASS' 'B-MAX' 'A4' 'KONA' 'I10' 'MOKKA'
 'S-MAX' 'X2' 'CROSSLAND X' 'TIGUAN' 'A5' 'GLE CLASS' 'ZAFIRA' 'IONIQ'
 'A6' 'MONDEO' 'YETI OUTDOOR' 'X1' 'SCALA' 'S CLASS' '1 SERIES' 'KAMIQ'
 'KUGA' 'TOURNEO CONNECT' 'Q7' 'GLA CLASS' 'ARTEON' 'SL CLASS' 'SANTA FE'
 'GRANDLAND X' 'I800' 'RAV4' 'TOURAN' 'CITIGO' 'ROOMSTER' 'PRIUS'
 'COROLLA' 'B CLASS' 'Q' 'KODIAQ' 'V CLASS' 'CADDY MAXI LIFE' 'SUPERB'
 'GETZ' 'COMBO LIFE' 'BEETLE' 'GALAXY' 'M3' 'GTC' 'X4' 'KA' 'IX35'
 'GRAND TOURNEO CONNECT' 'M4' 'TOURNEO CUSTOM' 'Z4

## 9.Numerical Features

In [2786]:
numerical_columns

Index(['Year', 'Price', 'Mileage', 'Tax', 'MPG', 'EngineSize', 'PaintQuality%',
       'PreviousOwners', 'Hasdamage'],
      dtype='object')

In [2787]:
for col in numerical_columns:
    print(f"\n--- {col} ---")
    print(df_training[col].unique())


--- Year ---
[2016.         2019.         2018.         2014.         2017.
 2020.         2013.                   nan 2015.         2023.36707842
 2011.         2012.         2023.1169636  2003.         2009.
 2007.         2005.         2011.11118842 2011.21085349 2023.97731126
 2004.         2010.         2010.56500919 2008.         2024.12175905
 2006.         2023.60527574 2023.38982198 2010.67696784 2001.
 2000.         2023.26798867 2010.26863473 2010.37154646 2002.
 2022.69668507 2009.81675711 2010.7464032  2012.69574039 1996.
 1998.         2022.87800554 1970.         1999.         1997.        ]

--- Price ---
[22290 13790 24990 ... 33606 16809 25785]

--- Mileage ---
[28421.  4589.  3624. ... 13613. 52134. 69072.]

--- Tax ---
[         nan 145.          30.          20.         150.
 160.         -56.19320154   0.         125.         200.
 250.         135.         300.         205.         260.
 326.02823317 325.         -45.78061049 265.         305.
 303.14889146 235. 

We can see that there are fetures that have negative numbers which doesn't make sense

In [2788]:
for col in numerical_columns:
    df_training.loc[df_training[col] < 0, col] = np.nan
for col in numerical_columns:
    if col!='Price':
        df_test.loc[df_test[col] < 0, col] = np.nan

Year and Previous Owners has float numbers wich should just have integer, so I will round all the numbers to have 0 decimal values

In [2789]:
df_training['Year']=df_training['Year'].round()
df_training['PreviousOwners']=df_training['PreviousOwners'].round()
df_test['Year']=df_test['Year'].round()
df_test['PreviousOwners']=df_test['PreviousOwners'].round()

To be easier to read and also to reduse noise we will round the rest to 2 decimals

In [2790]:
col_remove=['Year', 'PreviousOwners']
new_numerical_columns=[i  for i in numerical_columns if i not in col_remove]
for col in new_numerical_columns:
    df_training[col]= df_training[col].round(2)
for col in new_numerical_columns:
    if col!="Price" :
        df_test[col]= df_test[col].round(2)

**PaintQuality%** values greater than 100% do not make sense in this context, because we assume that 100% represents perfect overall paint quality and hull integrity of the car

In [2791]:
df_training.loc[df_training['PaintQuality%'] > 100] = np.nan
df_test.loc[df_test['PaintQuality%'] > 100]  = np.nan

In [2792]:
inconsistent_col = numerical_columns

for col in inconsistent_col:
    print(f'---{col}---')
    print(df_training[col][~df_training[col].apply(lambda x: isinstance(x, (int, float)))])


---Year---
Series([], Name: Year, dtype: float64)
---Price---
Series([], Name: Price, dtype: float64)
---Mileage---
Series([], Name: Mileage, dtype: float64)
---Tax---
Series([], Name: Tax, dtype: float64)
---MPG---
Series([], Name: MPG, dtype: float64)
---EngineSize---
Series([], Name: EngineSize, dtype: float64)
---PaintQuality%---
Series([], Name: PaintQuality%, dtype: float64)
---PreviousOwners---
Series([], Name: PreviousOwners, dtype: float64)
---Hasdamage---
Series([], Name: Hasdamage, dtype: float64)


In [2793]:
df_training.describe(include='all').T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
Brand,74092,9,FORD,15981,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Model,74099,204,FOCUS,6883,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Year,74121.0,NaN,NaN,NaN,2017.097489,2.204756,1970.0,2016.0,2017.0,2019.0,2024.0
Price,75606.0,NaN,NaN,NaN,16885.736449,9741.237162,450.0,10200.0,14699.0,20950.0,159999.0
Transmission,74090,4,MANUAL,41422,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Mileage,73784.0,NaN,NaN,NaN,23344.618425,21608.007007,1.0,7552.75,17411.5,32500.0,323000.0
FuelType,74103,5,PETROL,40986,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Tax,67369.0,NaN,NaN,NaN,121.259141,64.503161,0.0,125.0,145.0,145.0,580.0
MPG,67691.0,NaN,NaN,NaN,55.206729,16.361457,1.1,46.3,54.3,62.8,470.8
EngineSize,74012.0,NaN,NaN,NaN,1.662157,0.570913,0.0,1.2,1.6,2.0,6.6


## 10. Duplicated Rows

In [2794]:
#check if there is duplicated rows
duplicates = df_training[df_training.duplicated(keep=False)]
sorted_duplicates = duplicates.sort_values(by=list(df_training.columns))

sorted_duplicates

,Brand,Model,Year,Price,Transmission,Mileage,FuelType,Tax,MPG,EngineSize,PaintQuality%,PreviousOwners,Hasdamage
23477,FORD,FOCUS,2017.0,14000.0,AUTOMATIC,9137.0,PETROL,NaN,NaN,1.0,48.0,4.0,0.0
25766,FORD,FOCUS,2017.0,14000.0,AUTOMATIC,9137.0,PETROL,NaN,NaN,1.0,48.0,4.0,0.0
15152,FORD,FOCUS,2018.0,13500.0,MANUAL,11382.0,PETROL,NaN,NaN,1.0,57.0,4.0,0.0
21404,FORD,FOCUS,2018.0,13500.0,MANUAL,11382.0,PETROL,NaN,NaN,1.0,57.0,4.0,0.0
28593,FORD,FOCUS,2019.0,19500.0,MANUAL,3000.0,DIESEL,NaN,NaN,1.5,91.0,4.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
75768,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
39666,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
23930,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
38756,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [2795]:
df_training = df_training.drop_duplicates(keep='first')
df_test = df_test.drop_duplicates(keep='first')
df_test.shape

(32400, 12)

## Fill Missing Values

### Rows with all values equals to NaN

In [2796]:
df_training.isna().sum()

Brand             1515
Model             1508
Year              1486
Price                1
Transmission      1517
Mileage           1822
FuelType          1504
Tax               8234
MPG               7912
EngineSize        1595
PaintQuality%     1525
PreviousOwners    1915
Hasdamage         1545
dtype: int64

In [2797]:
rows_with_all_nan = df_training.isna().all(axis=1)

num_rows = rows_with_all_nan.sum()
print(f"Exists {num_rows} row/s completely none.")

if num_rows > 0:
    print(df_training[rows_with_all_nan])


Exists 1 row/s completely none.
      Brand Model  Year  Price Transmission  Mileage FuelType  Tax  MPG  \
15717   NaN   NaN   NaN    NaN          NaN      NaN      NaN  NaN  NaN   

       EngineSize  PaintQuality%  PreviousOwners  Hasdamage  
15717         NaN            NaN             NaN        NaN  


In [2798]:
df_training = df_training.dropna(how='all')

### Brand

Firstly, let's see the NaN **Brands Models**, so we can fill them based on their model.

In [2799]:
brand_missing = df_training[df_training['Brand'].isna() & df_training['Model'].notna()]
brand_missing

,Brand,Model,Year,Price,Transmission,Mileage,FuelType,Tax,MPG,EngineSize,PaintQuality%,PreviousOwners,Hasdamage
66591,NaN,T-ROC,2018.0,23000.0,SEMI-AUTOMATIC,584.0,PETROL,150.0,41.5,2.0,79.0,1.0,0.0
5034,NaN,A3,2016.0,11871.0,MANUAL,52519.0,DIESEL,20.0,72.4,1.6,60.0,2.0,0.0
33153,NaN,I20,2020.0,13966.0,MANUAL,3500.0,PETROL,145.0,47.9,1.0,41.0,2.0,0.0
45338,NaN,CL CLASS,2020.0,54995.0,SEMI-AUTOMATIC,NaN,PETROL,145.0,33.2,2.0,61.0,4.0,0.0
66712,NaN,POLO,2016.0,8650.0,MANUAL,32800.0,PETROL,20.0,60.1,1.0,54.0,4.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
72615,NaN,GOLF,2015.0,11750.0,MANUAL,79000.0,DIESEL,20.0,67.3,2.0,68.0,0.0,0.0
3436,NaN,Q5,2015.0,21995.0,SEMI-AUTOMATIC,18188.0,DIESEL,200.0,47.1,2.0,84.0,3.0,0.0
7392,NaN,A5,2019.0,31490.0,AUTOMATIC,10.0,PETROL,145.0,39.8,2.0,70.0,1.0,0.0
13545,NaN,5 SERIES,2019.0,26380.0,SEMI-AUTOMATIC,7609.0,DIESEL,145.0,65.7,2.0,67.0,3.0,0.0


Let's check if we have for each **Model**, multiples **Brand**.

In [2800]:
model_brand_counts = (
    df_training.groupby('Model')['Brand']
    .nunique()  
    .reset_index()
    .rename(columns={'Brand': 'Unique_Brands'})
)
models_with_multiple_brands = model_brand_counts[model_brand_counts['Unique_Brands'] > 1]
print(models_with_multiple_brands)

        Model  Unique_Brands
0    1 SERIES              2
1    2 SERIES              2
2    3 SERIES              2
3    4 SERIES              2
4    5 SERIES              2
5    6 SERIES              2
6    7 SERIES              2
7    8 SERIES              2
114        M2              2
115        M3              2
116        M4              2
189        X1              2
190        X2              2
191        X3              2
193        X5              2
194        X6              2
195        X7              2
201        Z4              2


Lets check each brands to see if there are mistakes

In [2801]:
for model in models_with_multiple_brands['Model']:
    brand=df_training[df_training['Model']==model]['Brand'].unique()
    print(f"Model: {model} has this brands: {brand}")

Model: 1 SERIES has this brands: ['BMW' 'VW' nan]
Model: 2 SERIES has this brands: ['BMW' 'VW' nan]
Model: 3 SERIES has this brands: ['BMW' nan 'VW']
Model: 4 SERIES has this brands: ['BMW' nan 'VW']
Model: 5 SERIES has this brands: ['VW' 'BMW' nan]
Model: 6 SERIES has this brands: ['BMW' 'VW' nan]
Model: 7 SERIES has this brands: ['BMW' 'VW' nan]
Model: 8 SERIES has this brands: ['BMW' 'VW']
Model: M2 has this brands: ['BMW' 'VW']
Model: M3 has this brands: ['BMW' 'VW' nan]
Model: M4 has this brands: ['BMW' 'VW']
Model: X1 has this brands: ['BMW' 'VW' nan]
Model: X2 has this brands: ['BMW' nan 'VW']
Model: X3 has this brands: ['BMW' 'VW' nan]
Model: X5 has this brands: ['BMW' nan 'VW']
Model: X6 has this brands: ['VW' 'BMW' nan]
Model: X7 has this brands: ['BMW' 'VW' nan]
Model: Z4 has this brands: ['BMW' 'VW' nan]


As we can see, all the models that have more than one brand, the brands are BMW and Volkswagen. However, all of these models actually belong only to BMW.

In [2802]:
#let's change the brand of the previous models from vw to bmw 
for model in models_with_multiple_brands['Model']:
    filter= (df_training['Model']==model) & (df_training['Brand']=='VW')
    df_training.loc[filter,'Brand']='BMW'

   

In [2803]:
#checking if there is no more brands that have nore then one brand
model_brand_counts = (
    df_training.groupby('Model')['Brand']
    .nunique()  
    .reset_index()
    .rename(columns={'Brand': 'Unique_Brands'})
)
models_with_multiple_brands = model_brand_counts[model_brand_counts['Unique_Brands'] > 1]
print(models_with_multiple_brands)

Empty DataFrame
Columns: [Model, Unique_Brands]
Index: []


In [2804]:
#make a dictionary of all brands based on the models (keys)
model_to_brand = df_training.groupby('Model')['Brand'].first().to_dict()
print(f"before(nº of missing values in brands): {df_training['Brand'].isna().sum()}")
# for each Brand it will fill the nan brands based on the model the car has(if the model is not nan)
df_training['Brand'] = df_training.apply(
    lambda row: model_to_brand[row['Model']] if (pd.isna(row['Brand']) and pd.notna(row['Model']))else row['Brand'],
    axis=1
)
print(f"after((nº of missing values in brands):  {df_training['Brand'].isna().sum()}")


before(nº of missing values in brands): 1514
after((nº of missing values in brands):  36


Now, the only remaining missing **Brand** values are those where the **Model** is also missing.

### Model

Since the model is more specific within a brand, we will determine it basedbased on the mode of the car's year, brand, fuel type, engine size, and average miles per gallon. 

In [2805]:
df_training = df_training.copy()

print(f"before(nº of missing values in modles):  {df_training['Model'].isna().sum()}")

brand_to_model = df_training.groupby( ['Brand','Year', 'FuelType', 'EngineSize', 'MPG'])['Model'] \
                            .agg(lambda x: x.mode().iloc[0] if not x.mode().empty else None) \
                            .to_dict()

df_training['Model'] = df_training.apply(
    lambda row: brand_to_model.get((row['Brand'], row['Year'],row['FuelType'],row['EngineSize'],row['MPG']), row['Model']) 
    if pd.isna(row['Model']) 
       and pd.notna(row['Brand']) 
       and pd.notna(row['Year']) 
       and pd.notna(row['FuelType']) 
       and pd.notna(row['EngineSize']) 
       and pd.notna(row['MPG'])
    else row['Model'],
    axis=1
)


print(f"after(nº of missing values in modles):  {df_training['Model'].isna().sum()}")


before(nº of missing values in modles):  1507
after(nº of missing values in modles):  349


We still have 349 rows where **Model** are missing values.



### FuelType

Since the same brand and model in the same year usally have the same Fueltype, we will fill the NaN values with the mode based on the brand, model and yaer

In [2806]:
df_training = df_training.copy()

print(f"before(nº of missing values in modles):  {df_training['FuelType'].isna().sum()}")

brand_to_model = df_training.groupby( ['Brand','Year', 'Model'])['FuelType'] \
                            .agg(lambda x: x.mode().iloc[0] if not x.mode().empty else None) \
                            .to_dict()

df_training['FuelType'] = df_training.apply(
    lambda row: brand_to_model.get((row['Brand'], row['Year'],row['Model']), row['FuelType']) 
    if pd.isna(row['FuelType']) 
       and pd.notna(row['Brand']) 
       and pd.notna(row['Year']) 
       and pd.notna(row['Model']) 
    else row['Model'],
    axis=1
)


print(f"Missing FuelType after filling: {df_training['FuelType'].isna().sum()}")


before(nº of missing values in modles):  1503
Missing FuelType after filling: 359


### For alll the categorical missing values

For the remaining categorical values, filling in Transmission is very ambiguous, as the same brand and model in the same year can have different types of Transmission, for example FORD Mustang (GT) in 2019 can be automatic and manual.

Similarly, filling the remaning NaN values of Brand, Model and FuelType would also be ambiguous, so we will replace these missing values with Unknown.

In [2808]:
categorical_cols = df_training.select_dtypes(include='object').columns

for col in categorical_cols:
    if df_training[col].isna:               
        df_training.loc[df_training[col].isna(),col]='UNKNOWN'
df_training.isna().sum()

Brand                0
Model                0
Year              1485
Price                0
Transmission         0
Mileage           1821
FuelType             0
Tax               8233
MPG               7911
EngineSize        1594
PaintQuality%     1524
PreviousOwners    1914
Hasdamage         1544
dtype: int64

### Numerical Features

For numerical features, we will use the median to fill in the missing values, based on the brand and model. We chose the median because unknown brands and models can have a wide range of dispersed values, and the median is less affected by outliers.

In [2814]:
df_training = df_training.copy()


numeric_cols = ['Year', 'Mileage', 'Tax', 'MPG', 'EngineSize', 'PaintQuality%', 'PreviousOwners', 'Hasdamage']


for col in numeric_cols:
    group_median = df_training.groupby(['Brand', 'Model'])[col].transform('mean')
    df_training[col] = df_training[col].fillna(group_median)
    #cause the Hyundai accent is the only car like that and has Mileage NaN
    brand_median = df_training.groupby('Brand')[col].transform('median')
    df_training[col] = df_training[col].fillna(brand_median)
    
 




### Categorical Features

For categorical features we will use the mode to fill missing values.

In [2812]:
df_training.isna().sum()

Brand             0
Model             0
Year              0
Price             0
Transmission      0
Mileage           0
FuelType          0
Tax               0
MPG               0
EngineSize        0
PaintQuality%     0
PreviousOwners    0
Hasdamage         0
dtype: int64

## 11. Export Dataset

In [2813]:
df_copy.to_excel("DataExplorationTrain.xlsx", index=False)


NameError: name 'df_copy' is not defined

In [ ]:
df_test_copy.to_excel("DataExplorationTest.xlsx", index=False)
